File: src/doc_update_tool/tracking/tracking_sheet.py
"""Safely add/update one row in the user's existing Excel tracking sheet.

Loads the workbook in place (openpyxl, not pandas) so existing rows, styles, and any
other sheets in the file are left untouched. Re-running for the same version_no updates
that row instead of appending a duplicate. Backs up before writing and restores on any
failure, since this file is the user's own record and must never end up corrupted or
partially written.
"""

from __future__ import annotations

import shutil
from copy import copy
from pathlib import Path

import openpyxl

from doc_update_tool.config import TrackingSheetConfig

_HYPERLINK_FIELDS = {"email_path", "bilingual_file_path", "diff_report_path"}


class TrackingSheetError(RuntimeError):
    pass


def upsert_row(config: TrackingSheetConfig, version_no: int, row_data: dict) -> None:
    path = config.path
    if not path.exists():
        raise TrackingSheetError(
            f"追蹤表不存在: {path}。請先建立好既有追蹤表（含表頭），本工具只會新增/更新資料列。"
        )

    backup_path = path.with_suffix(path.suffix + ".bak")
    shutil.copy2(path, backup_path)

    try:
        wb = openpyxl.load_workbook(path)
        ws = wb[config.sheet_name]

        version_col = config.columns["version_no"]
        target_row = _find_row_by_version(ws, version_col, config.header_row, version_no)

        if target_row is None:
            target_row = ws.max_row + 1
            if ws.max_row > config.header_row:
                _copy_row_style(ws, ws.max_row, target_row)

        for field, value in row_data.items():
            col_letter = config.columns.get(field)
            if col_letter is None:
                continue  # unrecognized field name, ignore rather than fail the whole upsert
            cell = ws[f"{col_letter}{target_row}"]
            if field in _HYPERLINK_FIELDS and value:
                value_path = Path(value)
                cell.value = value_path.name
                cell.hyperlink = str(value_path)
            else:
                cell.value = value

        wb.save(path)
        openpyxl.load_workbook(path)  # sanity check: the file we just wrote must still open
    except Exception:
        shutil.copy2(backup_path, path)
        raise
    finally:
        backup_path.unlink(missing_ok=True)


def read_row(config: TrackingSheetConfig, version_no: int) -> dict[str, object] | None:
    if not config.path.exists():
        return None
    wb = openpyxl.load_workbook(config.path, data_only=True)
    ws = wb[config.sheet_name]
    row = _find_row_by_version(ws, config.columns["version_no"], config.header_row, version_no)
    if row is None:
        return None
    return {field: ws[f"{col}{row}"].value for field, col in config.columns.items()}


def _find_row_by_version(ws, version_col: str, header_row: int, version_no: int) -> int | None:
    for row in range(header_row + 1, ws.max_row + 1):
        cell_value = ws[f"{version_col}{row}"].value
        if cell_value is not None and str(cell_value).strip() == str(version_no):
            return row
    return None


def _copy_row_style(ws, src_row: int, dst_row: int) -> None:
    for col_idx in range(1, ws.max_column + 1):
        src_cell = ws.cell(row=src_row, column=col_idx)
        dst_cell = ws.cell(row=dst_row, column=col_idx)
        dst_cell.font = copy(src_cell.font)
        dst_cell.border = copy(src_cell.border)
        dst_cell.fill = copy(src_cell.fill)
        dst_cell.number_format = src_cell.number_format
        dst_cell.alignment = copy(src_cell.alignment)

File: web/src/tracking-sheet.ts
/** Mirrors tracking/tracking_sheet.py's safe upsert — but adapted for the browser:
 * there's no real filesystem path to hyperlink to (Path B has no folder access), so
 * path-like fields are written as plain descriptive text (e.g. the name the file will
 * have once the user unzips the output bundle) rather than a clickable hyperlink.
 *
 * There's also no on-disk backup/restore step: the caller (app.ts) keeps the original
 * uploaded bytes until upsertRow resolves successfully, so a mid-way failure here just
 * means the original tracking sheet file is never touched/downloaded.
 */

import type ExcelJSNS from "exceljs";

import { TRACKING_FIELD_LABELS } from "./config.ts";
import type { TrackingColumns, TrackingSheetMapping } from "./config.ts";
import { toArrayBuffer } from "./xlsx-buffer.ts";

export type TrackingRowData = Partial<Record<keyof TrackingSheetMapping["columns"], string | number>>;

export class TrackingSheetError extends Error {}

export async function upsertRow(
  ExcelJS: typeof ExcelJSNS,
  bytes: ArrayBuffer,
  mapping: TrackingSheetMapping,
  versionNo: number,
  rowData: TrackingRowData
): Promise<ArrayBuffer> {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(bytes);
  const ws = wb.getWorksheet(mapping.sheetName);
  if (!ws) throw new TrackingSheetError(`追蹤表內找不到工作表: ${mapping.sheetName}`);

  let targetRow = findRowByVersion(ws, mapping, versionNo, rowData.sourceFileName as string | undefined);

  if (targetRow === null) {
    targetRow = ws.rowCount + 1;
    if (ws.rowCount > mapping.headerRow) copyRowStyle(ws, ws.rowCount, targetRow);
  }

  for (const [field, value] of Object.entries(rowData)) {
    const colLetter = mapping.columns[field as keyof TrackingSheetMapping["columns"]];
    if (!colLetter) continue; // unrecognized field name, ignore rather than fail the whole upsert
    ws.getCell(`${colLetter}${targetRow}`).value = value as string | number;
  }

  return toArrayBuffer(await wb.xlsx.writeBuffer());
}

export async function readRow(
  ExcelJS: typeof ExcelJSNS,
  bytes: ArrayBuffer,
  mapping: TrackingSheetMapping,
  versionNo: number
): Promise<Record<string, unknown> | null> {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(bytes);
  const ws = wb.getWorksheet(mapping.sheetName);
  if (!ws) return null;

  const row = findRowByVersion(ws, mapping, versionNo);
  if (row === null) return null;

  const result: Record<string, unknown> = {};
  for (const [field, col] of Object.entries(mapping.columns)) {
    result[field] = ws.getCell(`${col}${row}`).value;
  }
  return result;
}

export interface TrackingEntry {
  versionNo: number;
  sourceFileName: string;
}

/** Lists (versionNo, sourceFileName) for every existing row, used to match a newly
 * uploaded file against tracking history by filename (see plan Context: customer
 * filenames often change slightly between versions, e.g. a date suffix, so this is a
 * hint for the user to confirm/pick from rather than something to silently trust). */
export async function listEntries(
  ExcelJS: typeof ExcelJSNS,
  bytes: ArrayBuffer,
  mapping: TrackingSheetMapping
): Promise<TrackingEntry[]> {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(bytes);
  const ws = wb.getWorksheet(mapping.sheetName);
  if (!ws) return [];

  const entries: TrackingEntry[] = [];
  for (let row = mapping.headerRow + 1; row <= ws.rowCount; row++) {
    const versionValue = ws.getCell(`${mapping.columns.versionNo}${row}`).value;
    if (versionValue == null || String(versionValue).trim() === "") continue;
    const sourceFileName = String(ws.getCell(`${mapping.columns.sourceFileName}${row}`).value ?? "");
    entries.push({ versionNo: Number(versionValue), sourceFileName });
  }
  return entries;
}

/** Matches on versionNo alone by default, but if `sourceFileName` is given, also
 * requires that column to match — a shared tracking sheet can hold several unrelated
 * documents' histories side by side (e.g. an Excel spec and a Word spec attached to the
 * same batch of files), and each one's version numbering starts independently at 1, so
 * versionNo alone isn't a safe row key once more than one document family is involved:
 * without this, upserting the second document's "version 1" would silently overwrite
 * the first document's "version 1" row instead of adding its own. */
function findRowByVersion(
  ws: ExcelJSNS.Worksheet,
  mapping: TrackingSheetMapping,
  versionNo: number,
  sourceFileName?: string
): number | null {
  for (let row = mapping.headerRow + 1; row <= ws.rowCount; row++) {
    const value = ws.getCell(`${mapping.columns.versionNo}${row}`).value;
    if (value == null || String(value).trim() !== String(versionNo)) continue;
    if (sourceFileName !== undefined) {
      const rowFileName = String(ws.getCell(`${mapping.columns.sourceFileName}${row}`).value ?? "");
      if (rowFileName !== sourceFileName) continue;
    }
    return row;
  }
  return null;
}

export type TrackingRow = Record<keyof TrackingColumns, string>;

const TRACKING_FIELD_KEYS = Object.keys(TRACKING_FIELD_LABELS) as Array<keyof TrackingColumns>;

/** Reads every data row's every column (unlike listEntries, which only reads
 * versionNo/sourceFileName for filename-matching) — powers the tracking-management
 * tab's editable table, where the user can see and change any field on screen. */
export async function listAllRows(
  ExcelJS: typeof ExcelJSNS,
  bytes: ArrayBuffer,
  mapping: TrackingSheetMapping
): Promise<TrackingRow[]> {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(bytes);
  const ws = wb.getWorksheet(mapping.sheetName);
  if (!ws) return [];

  const rows: TrackingRow[] = [];
  for (let row = mapping.headerRow + 1; row <= ws.rowCount; row++) {
    const versionValue = ws.getCell(`${mapping.columns.versionNo}${row}`).value;
    const rowValues = TRACKING_FIELD_KEYS.map(
      (field) => ws.getCell(`${mapping.columns[field]}${row}`).value
    );
    const rowIsBlank = versionValue == null && rowValues.every((v) => v == null || String(v).trim() === "");
    if (rowIsBlank) continue;

    const result = {} as TrackingRow;
    for (let i = 0; i < TRACKING_FIELD_KEYS.length; i++) {
      result[TRACKING_FIELD_KEYS[i]] = rowValues[i] == null ? "" : String(rowValues[i]);
    }
    rows.push(result);
  }
  return rows;
}

/** Rebuilds the tracking sheet's data rows from the given in-memory rows (the state of
 * the on-screen editable table) — used both to save edits back onto an uploaded tracking
 * file and to create a brand-new one from scratch when the user has none yet. Existing
 * data rows are cleared and replaced wholesale rather than diffed cell-by-cell: this is
 * a "what you see is what gets saved" table editor, not a merge. */
export async function buildTrackingWorkbook(
  ExcelJS: typeof ExcelJSNS,
  mapping: TrackingSheetMapping,
  rows: TrackingRow[],
  existingBytes?: ArrayBuffer
): Promise<ArrayBuffer> {
  const wb = new ExcelJS.Workbook();
  let ws: ExcelJSNS.Worksheet;
  let templateRow: number | null = null;

  if (existingBytes) {
    await wb.xlsx.load(existingBytes);
    const existing = wb.getWorksheet(mapping.sheetName);
    if (existing) {
      ws = existing;
      if (ws.rowCount > mapping.headerRow) templateRow = mapping.headerRow + 1;
      const clearThrough = Math.max(ws.rowCount, mapping.headerRow + rows.length);
      for (let row = mapping.headerRow + 1; row <= clearThrough; row++) {
        for (const field of TRACKING_FIELD_KEYS) {
          ws.getCell(`${mapping.columns[field]}${row}`).value = null;
        }
      }
    } else {
      ws = wb.addWorksheet(mapping.sheetName);
      writeHeaderRow(ws, mapping);
    }
  } else {
    ws = wb.addWorksheet(mapping.sheetName);
    writeHeaderRow(ws, mapping);
  }

  rows.forEach((rowData, i) => {
    const targetRow = mapping.headerRow + 1 + i;
    if (templateRow !== null && targetRow !== templateRow) copyRowStyle(ws, templateRow, targetRow);
    for (const field of TRACKING_FIELD_KEYS) {
      const value = rowData[field];
      ws.getCell(`${mapping.columns[field]}${targetRow}`).value = value === "" ? null : value;
    }
  });

  return toArrayBuffer(await wb.xlsx.writeBuffer());
}

function writeHeaderRow(ws: ExcelJSNS.Worksheet, mapping: TrackingSheetMapping): void {
  for (const field of TRACKING_FIELD_KEYS) {
    ws.getCell(`${mapping.columns[field]}${mapping.headerRow}`).value = TRACKING_FIELD_LABELS[field];
  }
}

function copyRowStyle(ws: ExcelJSNS.Worksheet, srcRow: number, dstRow: number): void {
  for (let col = 1; col <= ws.columnCount; col++) {
    const srcCell = ws.getRow(srcRow).getCell(col);
    const dstCell = ws.getRow(dstRow).getCell(col);
    // Deep-clone before assigning: after a save/reload round trip, ExcelJS cells with
    // identical style values share the same in-memory style object, and a later plain
    // `cell.font = {...}` mutates that shared object in place rather than detaching
    // first — so any code that mutates an *existing* row's style after creation (not
    // done anywhere today) must clone first or it will silently reach into other rows
    // with the same style.
    dstCell.style = JSON.parse(JSON.stringify(srcCell.style));
  }
}
